samza-operator/src/main/java/org/apache/samza/operators/StreamOperatorAdaptorTask.java

```diff
@@ -0,0 +1,107 @@
+ * transformation chains.
+ * <p>
+ * It accepts an instance of the user implemented {@link StreamOperatorTask}. When its own {@link #init(Config, TaskContext)}
+ * method is called during startup, it creates a {@link MessageStreamImpl} corresponding to each of its input
+ * {@link SystemStreamPartition}s and then calls the user's {@link StreamOperatorTask#transform(Map)} method.
+ * <p>
+ * When users invoke the methods on the {@link MessageStream} API to describe their stream transformations in the
+ * {@link StreamOperatorTask#transform(Map)} method, the underlying {@link MessageStreamImpl} creates the
+ * corresponding {@link org.apache.samza.operators.spec.OperatorSpec} to record information about the desired
+ * transformation, and returns the output {@link MessageStream} to allow further transform chaining.
+ * <p>
+ * Once the user's transformation DAGs have been described for all {@link MessageStream}s (i.e., when the
+ * {@link StreamOperatorTask#transform(Map)} call returns), it calls
+ * {@link OperatorImpls#createOperatorImpls(MessageStreamImpl, TaskContext)} for each of the input
+ * {@link MessageStreamImpl}. This instantiates the {@link org.apache.samza.operators.impl.OperatorImpl} DAG
+ * corresponding to the aforementioned {@link org.apache.samza.operators.spec.OperatorSpec} DAG and returns the
+ * root node of the DAG, which this class saves.
+ * <p>
+ * Now that it has the root for the DAG corresponding to each {@link SystemStreamPartition}, it can pass the messages
+ * received in {@link StreamTask#process(IncomingMessageEnvelope, MessageCollector, TaskCoordinator)} along to the
+ * appropriate root nodes. From then on, each {@link org.apache.samza.operators.impl.OperatorImpl} propagates its
+ * transformed output to the next set of {@link org.apache.samza.operators.impl.OperatorImpl}s.
+ */
+public final class StreamOperatorAdaptorTask implements StreamTask, InitableTask, WindowableTask {
+
+  /**
+   * A mapping from each {@link SystemStreamPartition} to the root node of its operator chain DAG.
+   */
+  private final Map<SystemStreamPartition, OperatorImpl<IncomingSystemMessage, ? extends Message>> operatorChains = new HashMap<>();
+
+  private final StreamOperatorTask userTask;
+
+  public StreamOperatorAdaptorTask(StreamOperatorTask userTask) {
+    this.userTask = userTask;
+  }
+
+  @Override
+  public final void init(Config config, TaskContext context) throws Exception {
+    if (this.userTask instanceof InitableTask) {
+      ((InitableTask) this.userTask).init(config, context);
+    }
+    Map<SystemStreamPartition, MessageStream<IncomingSystemMessage>> messageStreams = new HashMap<>();
+    context.getSystemStreamPartitions().forEach(ssp -> messageStreams.put(ssp, new MessageStreamImpl<>()));
+    this.userTask.transform(messageStreams);
+    messageStreams.forEach((ssp, ms) -> {
+        OperatorImpl<IncomingSystemMessage, ? extends Message> rootOperator =
```

Comment:
minor, it would be simpler like the following:
operatorChains.put(ssp, OperatorImpls.createOperatorImpls((MessageStreamImpl<IncomingSystemMessage>) ms, context));